### PR TITLE
feat(drive): logging to multiple sources

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

We need to be able to log to multiple destinations, like files and standard out. 

The following requirements need to be met:

1. Support buffering and flush to reduce logging overhead
2. Support different logging levels
3. Ability to write into stdout and files with different logging levels
4. Abilty to write JSON or human readable format
5. Ability to log some data not just message

## What was done?

1. Support buffering and flush to reduce logging overhead - yes, buffering handled inside [tracing_subscriber](https://github.com/tokio-rs/tracing/blob/29146260fb4615d271d2e899ad95a753bb42915e/tracing-subscriber/src/fmt/fmt_subscriber.rs#L934) crate
2. Support different logging levels - done, using [ABCI_LOG_<key>_VERBOSITY](https://github.com/dashpay/platform/blob/b0d041e86f2520ec8e4986eb8cd0e59fdd95c378/packages/rs-drive-abci/src/logging.rs#L40-L41)
3. Ability to write into stdout and files with different logging levels - done, using [ABCI_LOG_<key>_DESTINATION](https://github.com/dashpay/platform/blob/b0d041e86f2520ec8e4986eb8cd0e59fdd95c378/packages/rs-drive-abci/src/logging.rs#L31-L39)
4. Abilty to write JSON or human readable format - done, using [ABCI_LOG_<key>_FORMAT=Json](https://github.com/dashpay/platform/blob/b0d041e86f2520ec8e4986eb8cd0e59fdd95c378/packages/rs-drive-abci/src/logging.rs#L45-L56)
5. Ability to log some data not just message - standard feature of tracing, see https://docs.rs/tracing/latest/tracing/#recording-fields

Example JSON logs:

```json
{"timestamp":"2023-05-22T13:21:32.226938Z","level":"TRACE","fields":{"message":"sending response","response":"VerifyVoteExtension(ResponseVerifyVoteExtension { status: Accept })"},"target":"tenderdash_abci::application"}
{"timestamp":"2023-05-22T13:21:32.279373Z","level":"TRACE","fields":{"message":"received request","request":"Request { value: Some(VerifyVoteExtension(RequestVerifyVoteExtension { hash: [105, 185, 250, 113, 158, 75, 42, 239, 172, 10, 167, 198, 58, 208, 171, 82, 149, 245, 30, 186, 163, 159, 127, 118, 52, 244, 41, 108, 8, 227, 222, 1], validator_pro_tx_hash: [252, 22, 137, 215, 197, 95, 148, 249, 48, 150, 238, 85, 36, 56, 220, 215, 17, 12, 236, 133, 9, 8, 60, 145, 97, 210, 161, 76, 16, 9, 37, 193], height: 484, round: 0, vote_extensions: [] })) }"},"target":"tenderdash_abci::application"}
{"timestamp":"2023-05-22T13:21:32.279689Z","level":"TRACE","fields":{"message":"sending response","response":"VerifyVoteExtension(ResponseVerifyVoteExtension { status: Accept })"},"target":"tenderdash_abci::application"}
{"timestamp":"2023-05-22T13:21:32.379858Z","level":"TRACE","fields":{"message":"received request","request":"Request { value: Some(VerifyVoteExtension(RequestVerifyVoteExtension { hash: [105, 185, 250, 113, 158, 75, 42, 239, 172, 10, 167, 198, 58, 208, 171, 82, 149, 245, 30, 186, 163, 159, 127, 118, 52, 244, 41, 108, 8, 227, 222, 1], validator_pro_tx_hash: [253, 81, 25, 148, 210, 112, 163, 44, 105, 7, 222, 39, 165, 177, 200, 94, 121, 243, 235, 36, 107, 121, 34, 98, 63, 34, 89, 218, 83, 238, 108, 166], height: 484, round: 0, vote_extensions: [] })) }"},"target":"tenderdash_abci::application"}
{"timestamp":"2023-05-22T13:21:32.380069Z","level":"TRACE","fields":{"message":"sending response","response":"VerifyVoteExtension(ResponseVerifyVoteExtension { status: Accept })"},"target":"tenderdash_abci::application"}
{"timestamp":"2023-05-22T13:21:32.408517Z","level":"TRACE","fields":{"message":"check if block info matches request","self":"BlockStateInfo { height: 484, round: 0, block_time_ms: 1684761691632, previous_block_time_ms: Some(1684761680079), proposer_pro_tx_hash: [57, 91, 68, 134, 95, 63, 126, 212, 30, 76, 67, 160, 176, 148, 85, 57, 3, 216, 202, 24, 231, 158, 228, 89, 7, 233, 144, 223, 244, 229, 142, 19], core_chain_locked_height: 1732, block_hash: Some([105, 185, 250, 113, 158, 75, 42, 239, 172, 10, 167, 198, 58, 208, 171, 82, 149, 245, 30, 186, 163, 159, 127, 118, 52, 244, 41, 108, 8, 227, 222, 1]), app_hash: Some([116, 186, 161, 20, 194, 61, 167, 63, 7, 50, 158, 62, 36, 223, 97, 170, 197, 159, 40, 26, 80, 169, 69, 158, 43, 198, 244, 210, 0, 167, 214, 61]) }","height":"484","round":"0","core_block_height":"1732","proposer_pro_tx_hash":"395b44865f3f7ed41e4c43a0b094553903d8ca18e79ee45907e990dff4e58e13","commit_hash":"69b9fa719e4b2aefac0aa7c63ad0ab5295f51ebaa39f7f7634f4296c08e3de01"},"target":"drive_abci::block"}

```

## How Has This Been Tested?

* started platform manually in docker
* added unit test

## Breaking Changes

None, changes are backwards-compatible.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/dashmate/docker-compose.platform.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ services:
     restart: unless-stopped
     volumes:
       - drive_abci_data:/var/lib/dash/rs-drive-abci
+        drive_abci_logs:/var/log/dash
```
Comment (strophy) [inline packages/dashmate/docker-compose.platform.yml:9]: Shouldn't it be like this?
```suggestion
      - drive_abci_logs:/var/log/dash
```

**Review by strophy — DISMISSED** [review]
Looks excellent to me, reviewed readme, output and compose file. Not sure about Rust stuff.

File: packages/rs-drive-abci/src/logging.rs [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+
+/// Name of logging configuration
+pub type LoggerID = String;
+
+/// Add and manage logging methods
+pub struct LogController {
+    loggers: HashMap<LoggerID, Logger>,
+}
+
+impl Default for LogController {
```
Comment (shumkov) [inline packages/rs-drive-abci/src/logging.rs:175]: You can go with drive `#[derive(Default)]`

File: packages/rs-drive-abci/src/logging.rs [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+    /// Duplicate config
+    #[error("duplicate log configuration name {0}")]
+    DuplicateConfigName(String),
+}
+
+/// Name of logging configuration
+pub type LoggerID = String;
+
+/// Add and manage logging methods
+pub struct LogController {
```
Comment (shumkov) [inline packages/rs-drive-abci/src/logging.rs:171]: It looks more like a mixed Builder pattern with the collection for logs. I would separate builder logic from the container, so builder will create a container.

File: packages/rs-drive-abci/src/logging.rs [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        for (name, config) in configs {
+            self.add(name, config)?;
+        }
+        Ok(())
+    }
+
+    /// Flush all loggers.
+    ///
+    /// In case of multiple errors, returns only last one.
+    pub fn flush(&self) -> Result<(), std::io::Error> {
```
Comment (shumkov) [inline packages/rs-drive-abci/src/logging.rs:210]: I didn't find where do we use it?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/dashmate/docker-compose.platform.yml (hunk 1 above)
Comment (shumkov) [reply packages/dashmate/docker-compose.platform.yml]: yeah, we can do that

**Review by shumkov — APPROVED** [review]
Good job 👍

**Review by shumkov — APPROVED** [review]

**Pull request merged**